lib/src/facts/linux/dmi_resolver.cc

```diff
@@ -14,22 +16,155 @@ namespace facter { namespace facts { namespace linux {
+            result.manufacturer         = read("/sys/class/dmi/id/sys_vendor");
+            result.product_name         = read("/sys/class/dmi/id/product_name");
+            result.serial_number        = read("/sys/class/dmi/id/product_serial");
+            result.uuid                 = read("/sys/class/dmi/id/product_uuid");
+            result.chassis_type         = to_chassis_description(read("/sys/class/dmi/id/chassis_type"));
+        } else {
+            LOG_DEBUG("/sys/class/dmi cannot be accessed: using dmidecode to query DMI information.");
+
+            int dmi_type = -1;
+            execution::each_line("dmidecode", [&](string& line) {
+                parse_dmidecode_output(result, line, dmi_type);
+                return true;
+            });
+        }
         return result;
     }
 
+    void dmi_resolver::parse_dmidecode_output(data& result, string& line, int& dmi_type)
+    {
+        static const boost::regex dmi_section_pattern("^Handle 0x.{4}, DMI type (\\d{1,3})");
+
+        // Stores the relevant sections; this is in order based on DMI type ID
+        // Ensure there's a trailing semicolon on each entry and keep in sync with the switch statement below
+        static vector<vector<string>> sections = {
+            {   // BIOS (0)
+                "vendor:",
+                "version:",
+                "release date:",
+            },
+            {   // System (1)
+                "manufacturer:",
+                "product:",
+                "product name:",
+                "serial number:",
+                "uuid:",
+            },
+            {   // Base Board (2)
+                "manufacturer:",
+                "product:",
+                "product name:",
+                "serial number:",
+                "asset tag:",
+            },
+            {   // Chassis (3)
+                "type:",
+                "chassis type:",
+                "asset tag:",
+            }
+        };
+
+        // Check for a section header
+        if (re_search(line, dmi_section_pattern, &dmi_type)) {
+            return;
+        }
+
+        // Check that we're in a relevant section
+        if (dmi_type < 0 || static_cast<size_t>(dmi_type) >= sections.size()) {
+            return;
+        }
+
+        // Trim leading whitespace
+        boost::trim_left(line);
+
+        // Find a matching header
+        auto const& headers = sections[dmi_type];
+        auto it = find_if(headers.begin(), headers.end(), [&](string const& header) {
+            return boost::istarts_with(line, header);
+        });
+        if (it == headers.end()) {
+            return;
+        }
+
+        // Get the value and trim it
+        string value = line.substr(it->size());
+        boost::trim(value);
+
+        // Calculate the index into the header vector
+        size_t index = it - headers.begin();
+
+        // Assign to the appropriate member
+        string* member = nullptr;
+        switch (dmi_type) {
```

Comment:
This is not my favorite switch statement, but the alternative would create a vector of sections of pairs of string and address of a variable in result. I can't decide which is worse.